Add guidance on how to migrate scala.Seq (#1326)
scala/docs.scala-lang: _overviews/core/collections-migration-213.md

  - Deprecated collections were removed (`MutableList`, `immutable.Stack`, others)
  - Parallel collections are now in a separate hierarchy in a [separate module](https://github.com/scala/scala-parallel-collections).
  - The `scala.jdk.StreamConverters` object provides extension methods to create (sequential or parallel) Java 8 streams for Scala collections.

## Tools for migrating and cross-building

The [scala-collection-compat](https://github.com/scala/scala-collection-compat) is a library released for 2.11, 2.12 and 2.13 that provides some of the new APIs from Scala 2.13 for the older versions. This simplifies cross-building projects.

The module also provides [migratrion rules](https://github.com/scala/scala-collection-compat#migration-tool) for [scalafix](https://scalacenter.github.io/scalafix/docs/users/installation.html) that can update a project's source code to work with the 2.13 collections library.

## scala.Seq migration

In Scala 2.13 `scala.Seq[+A]` is an alias for `scala.collection.immutable.Seq[A]`, instead of `scala.collection.Seq[A]`. This change requires some planning depending on how your code is going to be used.

> If you're making an application, and simply migrating Scala 2.12 code base to 2.13, it might be ok to keep using `scala.Seq` in your code.

Reviewer: ```suggestion
If you're making an application, and simply migrating a Scala 2.12 code base to 2.13, it might be ok to keep using `scala.Seq` in your code.
```

> If you're making an application, and simply migrating a Scala 2.12 code base to 2.13, it might be ok to keep using `scala.Seq` in your code.

Reviewer 2: This sounds like "keep using `scala.Seq`" should be the safe and simple option. I'd argue it's the opposite (unless your code happens to just work out of the box). In a complex codebase the simplest way forward is to change all references from `scala.Seq` to `scala.collection.Seq` and possibly insert `toSeq` calls where necessary. This is easy to do and will cross-compile on 2.12 without any changes in semantics.

Switching your own codebase from `scala.collection.Seq` to `scala.immutable.Seq` is the more advanced refactoring. In many cases it may have been the intention all along, so it makes sense to do it. It also avoids unnecessary copying when varargs are involved, but it's much harder to get right consistently in a large codebase.

Author: Interesting. I'll mention recommendation to use scala.collection.Seq first then.

Author: I've added this in 2b4de2d.

> If you're making a library intended to be used by other programmers, then using `scala.Seq` or vararg is going to be a breaking change in the API semantics. For example, if there was a function `def orderFood(order: Seq[Order]): Seq[Food]`, previously the library user would have been able to pass in an array of `Order`, but it won't work for 2.13.

Reviewer: ```suggestion
If you're making a library intended to be used by other programmers, then using `scala.Seq` or varargs is going to be a breaking change in the API semantics. For example, if there was a function `def orderFood(order: Seq[Order]): Seq[Food]`, previously the library user would have been able to pass in an array of `Order`, but it won't work for 2.13.
```

- if you cross build with Scala 2.12 and want to maintain the API semantics for 2.13 version of your library, or

> - if your library users frequently uses mutable collections such as `Array`

Reviewer 2: If code has been optimized to the point that you build an array instead of a regular collection, you're probably okay calling `unsafeWrapArray`, so it will still work with immutable varargs.

Author: Arrays might come up simply as Java API that you provide as a subset of API.

> you can import `scala.collection.Seq` ("CSeq") explicitly in your code.

Reviewer 3: It might be worth mentioning that this is a potentially source-breaking change. If a user in 2.12 has:

```
// `Seq` is the default `scala.Seq`
val a: Seq[Food] = orderFood(myOrder)
```

Then that will no longer compile in 2.13

Reviewer 3: Also, it might be a good strategy for APIs to only move to `immutable.Seq` in the result type, but keep accepting `collection.Seq` arguments, if the method implementation can work fine with mutable collections,.

Author: I've added this in 2b4de2d.

~~~ scala
import scala.collection.Seq
~~~

> In the future when your API is able to break the source compatibility, it might also make sense to migrate towards the `scala.collection.immutable.Seq` ("ISeq") for both Scala 2.12 and Scala 2.13.

Reviewer 4: it doesn't seem like you refer to "ISeq" or "CSeq" anywhere, so the parentheticals can probably be removed

Author: ok.

Author: Added commits that ended up adding references to ISeq and CSeq.

~~~ scala
import scala.collection.immutable.Seq
~~~

> Note that in Scala 2.13 the sequence passed into a vararg as `orderFood(xs: _*)` must also be immutable. This is because the sequence passed into a vararg must conform to`scala.Seq` acoording to [Specification](https://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#function-applications). Thus if your API exposes varargs, it will be an unavoidable breaking change. This might affect Java interoperability.

Reviewer: ```suggestion
Note that in Scala 2.13 the sequence passed into as a varargs as `orderFood(xs: _*)` must also be immutable. This is because the sequence passed in as a varargs must conform to `scala.Seq` according to [SLS 6.6](https://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#function-applications). Thus, if your API exposes varargs it will be an unavoidable breaking change. This might affect Java interoperability.
```

## scala.Seq and scala.IndexedSeq migration

In Scala 2.13 `scala.Seq[+A]` is an alias for `scala.collection.immutable.Seq[A]` ("ISeq"), instead of `scala.collection.Seq[A]` ("CSeq"). Similarly, `scala.IndexedSeq[+A]` is an alias for `scala.collection.immutable.IndexedSeq[A]`. These changes require some planning depending on how your code is going to be used.

If you're making a library intended to be used by other programmers, then using `scala.Seq`, `scala.IndexedSeq`, or vararg is going to be a breaking change in the API semantics. For example, if there was a function `def orderFood(order: Seq[Order]): Seq[Food]`, previously the library user would have been able to pass in an array of `Order`, but it won't work for 2.13.

- if you cross build with Scala 2.12 and want to maintain the API semantics for 2.13 version of your library, or
- if your library users frequently uses mutable collections such as `Array`

> you can import collection Seq explicitly in your code.

Reviewer: "import collection Seq" looks weird and kind of ambiguous.

~~~

Another workaround might be to accept CSeq, but return ISeq.

~~~ scala
import scala.collection.{ Seq => CSeq }
import scala.collection.immutable.{ Seq => ISeq }

object FoodToGo {
  def orderFood(order: CSeq[Order]): ISeq[Food]
}
~~~

In the future when your API is able to break the source compatibility, it might also make sense to migrate towards ISeq for both Scala 2.12 and Scala 2.13.

~~~ scala
import scala.collection.immutable.{ Seq => ISeq }

object FoodToGo {
  def orderFood(order: ISeq[Order]): ISeq[Food]
}
~~~

Similarly, if you're making an end-user application, unifying to CSeq might be the easier and safer initial path especially for a larger and complex code base. Switching to ISeq will be a more advanced refactoring.

Note that in Scala 2.13 the sequence passed into as a varargs as `orderFood(xs: _*)` must also be immutable. This is because the sequence passed in as a varargs must conform to `scala.Seq` according to [SLS 6.6](https://www.scala-lang.org/files/archive/spec/2.12/06-expressions.html#function-applications). Thus, if your API exposes varargs it will be an unavoidable breaking change. This might affect Java interoperability.

### Masking scala.Seq

> To use the compiler to bad the use of plain `Seq`, you can declare your own `Seq` to mask `scala.Seq`.

Reviewer: s/bad/ban/
